File: app/src/main/java/io/github/ovso/massage/main/f_theme/ThemeFragment.java
package io.github.ovso.massage.main.f_theme;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.database.FirebaseDatabase;

import org.jetbrains.annotations.NotNull;

import butterknife.BindView;
import io.github.ovso.massage.R;
import io.github.ovso.massage.framework.Constants;
import io.github.ovso.massage.framework.SelectableItem;
import io.github.ovso.massage.framework.customview.BaseFragment;
import io.github.ovso.massage.framework.listener.OnCustomRecyclerItemClickListener;
import io.github.ovso.massage.main.base.WebviewAlertDialog;
import io.github.ovso.massage.main.f_theme.adapter.ThemeAdapter;
import io.github.ovso.massage.main.f_theme.adapter.ThemeAdapterView;
import io.github.ovso.massage.main.f_theme.model.Theme;
import io.github.ovso.massage.view.ui.player.LandscapeVideoActivity;
import io.github.ovso.massage.view.ui.player.PortraitVideoActivity;
import io.reactivex.disposables.CompositeDisposable;
import jp.wasabeef.recyclerview.animators.SlideInDownAnimator;

public class ThemeFragment extends BaseFragment
        implements ThemePresenter.View, OnCustomRecyclerItemClickListener<SelectableItem<Theme>> {

    @BindView(R.id.recyclerview)
    RecyclerView recyclerView;
    @BindView(R.id.root_view)
    View rootView;
    @BindView(R.id.progressbar)
    ProgressBar progressBar;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final ThemeAdapter adapter = new ThemeAdapter();
    ThemeAdapterView adapterView;
    private ThemePresenter presenter;

    @Override
    protected int getLayoutResID() {
        return R.layout.fragment_theme;
    }

    @Override
    protected void onActivityCreate(Bundle savedInstanceState) {
        adapter.compositeDisposable = compositeDisposable;
        presenter = new ThemePresenterImpl(
                this,
                adapter,
                FirebaseDatabase.getInstance().getReference(),
                compositeDisposable
        );
        presenter.onActivityCreate();
    }

    public static ThemeFragment newInstance() {
        return new ThemeFragment();
    }

    @Override
    public void showYoutubeUseWarningDialog() {
        new AlertDialog.Builder(getActivity()).setMessage(R.string.youtube_use_warning)
                .setPositiveButton(android.R.string.ok, null)
                .show();
    }

    @Override
    public void showLandscapeVideo(String videoId) {
        Intent intent = new Intent(getContext(), LandscapeVideoActivity.class);
        intent.putExtra("video_id", videoId);
        startActivity(intent);
    }

    @Override
    public void showVideoTypeDialog(DialogInterface.OnClickListener $onClickListener) {
        final DialogInterface.OnClickListener onClickListener =
                $onClickListener::onClick;
        new AlertDialog.Builder(getContext()).setMessage(R.string.please_select_the_player_mode)
                .setPositiveButton(R.string.portrait_mode,
                        onClickListener)
                .setNeutralButton(R.string.landscape_mode, onClickListener)
                .setNegativeButton(android.R.string.cancel, onClickListener)
                .show();
    }

    @Override
    public void setRecyclerView() {
        if (recyclerView.getItemAnimator() != null) {
            recyclerView.getItemAnimator().setChangeDuration(Constants.DURATION_RECYCLERVIEW_ANI);
            recyclerView.getItemAnimator().setRemoveDuration(Constants.DURATION_RECYCLERVIEW_ANI);
        }
        recyclerView.setItemAnimator(new SlideInDownAnimator());
        recyclerView.addItemDecoration(getRvDivider());
        recyclerView.setAdapter(adapter);
    }

    @NotNull
    private DividerItemDecoration getRvDivider() {
        final DividerItemDecoration divider = new DividerItemDecoration(requireContext(), RecyclerView.VERTICAL);
        final Drawable drawable = ContextCompat.getDrawable(requireContext(), R.drawable.all_rv_divider);
        if (drawable != null) {
            divider.setDrawable(drawable);
        }
        return divider;
    }

    @Override
    public void showMessage(int resId) {
        Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(String msg) {
        Toast.makeText(getContext(), msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void refreshRemove(int position) {
        adapterView.refreshRemove(position);
    }

    @Override
    public void showLoading() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void refresh() {
        adapter.refresh();
    }

    @Override
    public void refresh(int position) {
        adapterView.refresh(position);
    }

    @Override
    public void showPortraitVideo(String videoId) {
        Intent intent = new Intent(getContext(), PortraitVideoActivity.class);
        intent.putExtra("video_id", videoId);
        startActivity(intent);
    }

    @Override
    public void showWebViewDialog(Theme item) {
        WebviewAlertDialog webviewAlertDialog = new WebviewAlertDialog();
        webviewAlertDialog.url = item.url;
        webviewAlertDialog.flag = item.flag;
        webviewAlertDialog.onCreateDialog(new Bundle()).show();
    }

    @Override
    public void removeRefresh() {
        adapterView.refreshRemove();
    }

    @Override
    public void onItemClick(SelectableItem<Theme> item) {
        presenter.onItemClick(item);
    }

    @Override
    public void onVideoClick(int position, SelectableItem<Theme> item) {
        presenter.onVideoClick(position, item);
    }

    @Override
    public void onDestroyView() {
        presenter.onDestroyView();
        compositeDisposable.clear();
        super.onDestroyView();
    }
}

File: app/src/main/java/io/github/ovso/massage/main/f_theme/ThemeFragment.kt
package io.github.ovso.massage.main.f_theme

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.database.FirebaseDatabase
import io.github.ovso.massage.R
import io.github.ovso.massage.framework.Constants
import io.github.ovso.massage.framework.SelectableItem
import io.github.ovso.massage.framework.customview.BaseFragment
import io.github.ovso.massage.framework.listener.OnCustomRecyclerItemClickListener
import io.github.ovso.massage.main.f_theme.adapter.ThemeAdapter
import io.github.ovso.massage.main.f_theme.adapter.ThemeAdapterView
import io.github.ovso.massage.main.f_theme.model.Theme
import io.github.ovso.massage.view.ui.player.LandscapeVideoActivity
import io.github.ovso.massage.view.ui.player.PortraitVideoActivity
import io.reactivex.disposables.CompositeDisposable
import jp.wasabeef.recyclerview.animators.SlideInDownAnimator
import kotlinx.android.synthetic.main.fragment_theme.*

class ThemeFragment : BaseFragment(), ThemePresenter.View,
    OnCustomRecyclerItemClickListener<SelectableItem<Theme?>?> {
    var progressBar: ProgressBar? = null
    private val compositeDisposable = CompositeDisposable()
    private val adapter = ThemeAdapter()
    var adapterView: ThemeAdapterView? = null
    private var presenter: ThemePresenter? = null
    override fun getLayoutResID(): Int {
        return R.layout.fragment_theme
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter.compositeDisposable = compositeDisposable
        presenter = ThemePresenterImpl(
            this,
            adapter,
            FirebaseDatabase.getInstance().reference,
            compositeDisposable
        )
        presenter?.onActivityCreate()

    }

    override fun showYoutubeUseWarningDialog() {
        AlertDialog.Builder(activity).setMessage(R.string.youtube_use_warning)
            .setPositiveButton(android.R.string.ok, null)
            .show()
    }

    override fun showLandscapeVideo(videoId: String) {
        val intent = Intent(context, LandscapeVideoActivity::class.java)
        intent.putExtra("video_id", videoId)
        startActivity(intent)
    }

    override fun showVideoTypeDialog(`$onClickListener`: DialogInterface.OnClickListener) {
        val onClickListener =
            DialogInterface.OnClickListener { dialog: DialogInterface?, which: Int ->
                `$onClickListener`.onClick(
                    dialog,
                    which
                )
            }
        AlertDialog.Builder(context).setMessage(R.string.please_select_the_player_mode)
            .setPositiveButton(
                R.string.portrait_mode,
                onClickListener
            )
            .setNeutralButton(R.string.landscape_mode, onClickListener)
            .setNegativeButton(android.R.string.cancel, onClickListener)
            .show()
    }

    override fun setRecyclerView() {
        if (recyclerview.itemAnimator != null) {
            recyclerview.itemAnimator!!.changeDuration =
                Constants.DURATION_RECYCLERVIEW_ANI.toLong()
            recyclerview.itemAnimator!!.removeDuration =
                Constants.DURATION_RECYCLERVIEW_ANI.toLong()
        }
        recyclerview.itemAnimator = SlideInDownAnimator()
        recyclerview.addItemDecoration(rvDivider)
        recyclerview.adapter = adapter
    }

    private val rvDivider: DividerItemDecoration
        get() {
            val divider = DividerItemDecoration(requireContext(), RecyclerView.VERTICAL)
            val drawable = ContextCompat.getDrawable(requireContext(), R.drawable.all_rv_divider)
            if (drawable != null) {
                divider.setDrawable(drawable)
            }
            return divider
        }

    override fun showMessage(resId: Int) {
        Toast.makeText(context, resId, Toast.LENGTH_SHORT).show()
    }

    override fun showMessage(msg: String) {
        Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
    }

    override fun refreshRemove(position: Int) {
        adapterView!!.refreshRemove(position)
    }

    override fun showLoading() {
        progressBar!!.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        progressBar!!.visibility = View.GONE
    }

    override fun refresh() {
        adapter.refresh()
    }

    override fun refresh(position: Int) {
        adapterView!!.refresh(position)
    }

    override fun showPortraitVideo(videoId: String) {
        val intent = Intent(context, PortraitVideoActivity::class.java)
        intent.putExtra("video_id", videoId)
        startActivity(intent)
    }

    override fun showWebViewDialog(item: Theme) {
    }

    override fun removeRefresh() {
        adapterView!!.refreshRemove()
    }

    override fun onItemClick(item: SelectableItem<Theme?>?) {
        presenter?.onItemClick(item)
    }

    override fun onVideoClick(position: Int, item: SelectableItem<Theme?>?) {
        presenter?.onVideoClick(position, item)
    }

    override fun onDestroyView() {
        presenter?.onDestroyView()
        compositeDisposable.clear()
        super.onDestroyView()
    }

    companion object {
        @JvmStatic
        fun newInstance(): ThemeFragment {
            return ThemeFragment()
        }
    }

}